### PR TITLE
ankiclient: add four new aggregate frequency markers

### PR DESCRIPTION
**Pull request by precondition**
This commit adds support for 4 new markers ([introduced in Yomitan](https://github.com/themoeway/yomitan/blob/659b837b24c04b0505c735ed16da5b63c902bab2/docs/anki-integration.md?plain=1#L42..L45) in https://github.com/themoeway/yomitan/pull/600):
1. {frequency-harmonic-rank} (Default: 9999999)
2. {frequency-harmonic-occurrence} (Default: 0)
3. {frequency-average-rank} (Default: 9999999)
4. {frequency-average-occurrence} (Default: 0)

These markers aggregate the term frequency information of multiple dictionaries into a single numeric value, [often used for sorting Anki cards by frequency](https://github.com/MarvNC/JP-Resources#sorting-mined-anki-cards-by-frequency).

### CODE REVIEW

File: src/anki/ankiclient.cpp [hunk 1]
```diff
@@ -1384,6 +1386,19 @@ void AnkiClient::buildCommonNote(
...
+
+        value.replace(REPLACE_FREQ_HARMONIC_OCCU, 
+                positiveIntToQString(frequencyHarmonic, 0));
+
+        value.replace(REPLACE_FREQ_AVERAGE_RANK,  
+                positiveIntToQString(frequencyAverage,  9999999));
+
+        value.replace(REPLACE_FREQ_AVERAGE_OCCU,  
+                positiveIntToQString(frequencyAverage,  0));
+
```
Comment (precondition) [inline src/anki/ankiclient.cpp:1401]: I was a bit conflicted on whether to repeat `frequencyXXX < 0 ? ["999999"|"0"] : QString::number(frequencyXXX);` four times or to consolidate that into a `positiveIntToQString` method.

Comment (ripose-jp) [reply src/anki/ankiclient.cpp:1401]: I think you made the right choice

File: src/anki/ankiclient.cpp [hunk 2]
```diff
@@ -1384,6 +1386,19 @@ void AnkiClient::buildCommonNote(
...
         value.replace(REPLACE_FREQUENCIES,  frequencies);
+
+        value.replace(REPLACE_FREQ_HARMONIC_RANK, 
+                positiveIntToQString(frequencyHarmonic, 9999999));
+
+        value.replace(REPLACE_FREQ_HARMONIC_OCCU, 
+                positiveIntToQString(frequencyHarmonic, 0));
+
+        value.replace(REPLACE_FREQ_AVERAGE_RANK,  
+                positiveIntToQString(frequencyAverage,  9999999));
```
Comment (ripose-jp) [inline src/anki/ankiclient.cpp:1397]: Same as above

File: src/anki/ankiclient.cpp [hunk 3]
```diff
@@ -1765,6 +1780,83 @@ QString AnkiClient::buildFrequencies(const QList<Frequency> &frequencies)
     return freqStr;
 }
 
+std::vector<int> AnkiClient::getFrequencyNumbers(
+        const QList<Frequency> &frequencies)
+{
+    QString previousDictionary;
+    std::vector<int> frequencyNumbers;
+
+    for (const auto& frequencyEntry : frequencies)
```
Comment (ripose-jp) [inline src/anki/ankiclient.cpp:1789]: The type is short enough here I don't think you should use `auto`. Also I prefer my `&` next to the variable.

**Review by ripose-jp — CHANGES_REQUESTED** [review]

**Comment by ripose-jp** [comment]
This needs to be fixed. I recommend you download Qt Creator rather than editing .ui files directly.

image

File: src/anki/ankiclient.cpp [hunk 4]
```diff
@@ -1384,6 +1386,19 @@ void AnkiClient::buildCommonNote(
         value.replace(REPLACE_CLOZE_PREFIX, clozePrefix);
         value.replace(REPLACE_CLOZE_SUFFIX, clozeSuffix);
         value.replace(REPLACE_FREQUENCIES,  frequencies);
+
+        value.replace(REPLACE_FREQ_HARMONIC_RANK, 
+                positiveIntToQString(frequencyHarmonic, 9999999));
```
[the remark this answers is not in the document]
Comment (precondition) [reply src/anki/ankiclient.cpp]: Fixed in https://github.com/ripose-jp/Memento/pull/238/commits/a92c0c0854dbf417536134dedbb79e2bc0812077

File: src/anki/ankiclient.cpp (hunk 2 above)
Comment (precondition) [reply src/anki/ankiclient.cpp]: Fixed in https://github.com/ripose-jp/Memento/pull/238/commits/a92c0c0854dbf417536134dedbb79e2bc0812077

File: src/anki/ankiclient.cpp (hunk 3 above)
Comment (precondition) [reply src/anki/ankiclient.cpp]: Fixed in https://github.com/ripose-jp/Memento/pull/238/commits/a92c0c0854dbf417536134dedbb79e2bc0812077

**Comment by precondition** [comment]
> This needs to be fixed. I recommend you download Qt Creator rather than editing .ui files directly.
> 
> image

Fixed in https://github.com/ripose-jp/Memento/pull/238/commits/4e1a9402c4590fb91e02dbf9210a40e5ce604119. For some reason, Qt Creator mangled the whole file, rendering the diff unreadable so I manually post-processed the XML file to format it in the same order as originally. Additionally, since the description labels have no word wrapping, I omitted the very last part of the descriptions listed on Yomitan's page.

**Comment by ripose-jp** [comment]
Looks good. My only request is that you squash all your changes into one commit. You can just `git reset HEAD~3`, recommit your changes, then force push.

I do have a quick question though. It looks like the only difference between rank and occurrence tags is the default value. Is that all it's supposed to be?

**Comment by precondition** [comment]
> I do have a quick question though. It looks like the only difference between rank and occurrence tags is the default value. Is that all it's supposed to be?

It is funny you say that because I wondered the same thing. When I asked the question in the `#tmw-dev` channel, they confirmed that this is how it is indeed supposed to be.

**Review by ripose-jp — APPROVED** [review]

**Comment by ripose-jp** [comment]
>A possible extension one can consider is to infer the frequency sorting mode

I think trusting the user to figure it out on their own is better than employing imperfect heuristics.

Everything looks good. Merging this.

**Pull request merged**